commands/audit/sca/pnpm/pnpm.go

```diff
@@ -0,0 +1,168 @@
+	if err != nil || (lockFileExists && pnpmDirExists) {
+		return
+	}
+	// Install is needed
+	log.Debug("Installing Pnpm project:", workingDir)
+	return getPnpmCmd(pnpmExecPath, workingDir, "install").GetCmd().Run()
+}
+
+// Run 'pnpm ls ...' command (project must be installed) and parse the returned result to create a dependencies map of.
+func calculateDependencies(executablePath, workingDir string, params utils.AuditParams) (dependencyTrees []*xrayUtils.GraphNode, uniqueDeps []string, err error) {
+	lsArgs := append([]string{"--depth", "Infinity", "--json", "--long"}, params.Args()...)
+	npmLsCmdContent, err := getPnpmCmd(executablePath, workingDir, "ls", lsArgs...).RunWithOutput()
+	if err != nil {
+		return
+	}
+	log.Debug("Pnpm ls command output:\n", string(npmLsCmdContent))
+	output := &[]pnpmLsProject{}
+	if err = json.Unmarshal(npmLsCmdContent, output); err != nil {
+		return
+	}
+	dependencyTrees, uniqueDeps = parsePnpmLSContent(*output)
+	return
+}
+
+func parsePnpmLSContent(projectInfo []pnpmLsProject) (dependencyTrees []*xrayUtils.GraphNode, uniqueDeps []string) {
+	uniqueDepsSet := datastructures.MakeSet[string]()
+	for _, project := range projectInfo {
+		// Parse the dependencies into Xray dependency tree format
+		dependencyTree, uniqueProjectDeps := coreXray.BuildXrayDependencyTree(createProjectDependenciesTree(project), getDependencyId(project.Name, project.Version))
+		// Add results
+		dependencyTrees = append(dependencyTrees, dependencyTree)
+		uniqueDepsSet.AddElements(maps.Keys(uniqueProjectDeps)...)
+	}
+	uniqueDeps = uniqueDepsSet.ToSlice()
+	return
+}
+
+func createProjectDependenciesTree(project pnpmLsProject) map[string]coreXray.DepTreeNode {
+	treeMap := make(map[string]coreXray.DepTreeNode)
+	directDependencies := []string{}
+	// Handle production-dependencies
+	for depName, dependency := range project.Dependencies {
+		directDependency := getDependencyId(depName, dependency.Version)
+		directDependencies = append(directDependencies, directDependency)
+		appendTransitiveDependencies(directDependency, dependency.Dependencies, treeMap)
+	}
+	// Handle dev-dependencies
+	for depName, dependency := range project.DevDependencies {
+		directDependency := getDependencyId(depName, dependency.Version)
+		directDependencies = append(directDependencies, directDependency)
+		appendTransitiveDependencies(directDependency, dependency.Dependencies, treeMap)
+	}
+	if len(directDependencies) > 0 {
+		treeMap[getDependencyId(project.Name, project.Version)] = coreXray.DepTreeNode{Children: directDependencies}
+	}
+	return treeMap
+}
+
+// Return npm://<name>:<version> of a dependency
+func getDependencyId(depName, version string) string {
+	return utils.NpmPackageTypeIdentifier + depName + ":" + version
+}
+
+func appendTransitiveDependencies(parent string, dependencies map[string]pnpmLsDependency, result map[string]coreXray.DepTreeNode) {
```

Comment:
Two considerations to ponder (unsure if necessary in this scenario) -
1. Loop detection: If a loop is identified, terminate the recursion.
2. Node appearance restriction: We impose a limit of 15 appearances for each dependency to prevent memory bloat.

Reply:
1. Pnpm `ls` returns a tree, so there are no loops there. (I think it handles loops, if not the 'ls' will not stop running).
2. We already impose the limit (it's 10 not 15) using: coreXray.BuildXrayDependencyTree at: https://github.com/jfrog/jfrog-cli-core/blob/master/utils/xray/xrayutils.go#L39